### ISSUE TITLE
Change video track (Quality) on HLS stream

### ISSUE THREAD

**Issue by endlessmind** [google/ExoPlayer#676]
When playing HLS streams (like twitch live streams), the player will adopt the the best possible quality, as expected. But I would like to be able to let the users select from available qualities, like in the YouTube app. (If the users don't need the best possible quality and/or have a limited data plans)

But I can't find any way to actually set the video quality manually.
How does the YouTube app accomplice this?

Will the "Steps toward full multi-track support."-commit to the dev-branch be of any use for this kind of functionality?

**Comment by ojw28** [comment]
Full multi-track support should make this easier, because once we have that it should be relatively easy to have the video renderer advertise multiple tracks. like: AUTO,1080,720,480 (or whatever).

YouTube use DASH, where track selection is supported (albeit not in a particularly easy way) using MultiTrackChunkSource. SmoothStreaming playbacks can also make use of MultiTrackChunkSource.

**Comment by crossle** [comment]
+1

**Comment by b95505017** [comment]
Currently I post a message into HlsChunkSource and change two things:
````java
adaptiveMode = ADAPTIVE_MODE.NONE;
selectedVariantIndex = index;
````

**Comment by endlessmind** [comment]
Can you provide more details on this? 
Also, where do you pull the variant-list from?

**Comment by crossle** [comment]
`player.getTrackCount(DemoPlayer.TYPE_VIDEO);` not return variant video

**Comment by b95505017** [comment]
Any development news or plans for this issue?

**Comment by Alviere** [comment]
Looking forward to this feature.

**Comment by ojw28** [comment]
It's not possible to select between the available variants in ExoPlayer. The demo app will also list the available variants and allows selection between them (during playback tap the screen and press the "Video" button to see the available variants, or the "Audio" button if your playing audio only).

**Comment by ojw28** [comment]
Note that this is on the dev branch (and will make it into the next release).

**Comment by b95505017** [comment]
Can't wait to try all of them! Thanks!

**Comment by crossle** [comment]
How to seamless switch? @ojw28

**Comment by ojw28** [comment]
We support adaptive mode where we'll seamlessly switch based on the available bandwidth, and we now also support manual selection. We do not support seamless manual selection.

**Comment by b95505017** [comment]
I tried the demo with some `live` playlist, and have several ideas:
1. Sometimes player stuck at buffering after `setSelectedTrack` for a long time, not sure it's network or player issue.
2. It would be nice without re-buffering if switch to `auto`.

**Comment by ojw28** [comment]
Thanks for your observations.

- I see the first issue, and we'll take a look and fix that.
- We don't have current plans to support the second It's pretty non-trivial. Consider the case where the player has buffered a large amount of some manually selected low quality stream. If auto is then selected, it may take so long to switch up that the selection appears to have not done anything. To mitigate that you can proactively throw away some of the buffer containing the low quality stream, but then things start getting complicated from an implementation perspective. It's also debatable how useful the feature would be. If someone has switched from auto to a manual selection then presumably they did so because auto wasn't working well for them. So presumably switching back to auto again is something few people would choose to do.
- We have a longer term plan to do something that achieves more or less the same thing as in your third bullet point. Basically, we'll at some point allow selection of a subset of streams that should be considered in auto mode.

**Comment by ojw28** [comment]
There's a separate issue tracking switching in Live (#1174), so let's use that for tracking.

**Comment by littlebob** [comment]
Hello @ojw28 
I see that seamless manual selection is not supported now. 
any plan to have this feature? 
If not, any idea for how one can implement this and creates a pull request? Thanks

**Comment by ojw28** [comment]
No immediate plans, no.

**Comment by ojw28** [comment]
In ExoPlayer V2, when it's done, HLS will use a `FormatEvaluator` to adapt between the variants during playback, as is already the case for DASH and SmoothStreaming. At that point you'll be able to implement this functionality using a custom `FormatEvaluator` implementation. For now in V1, you'd need to look at modifying HlsChunkSource.

**Comment by natez0r** [comment]
@ojw28 any idea when we can expect to be able to use `FormatEvaluator` for HLS? Is there a timetable for V2?

**Comment by toanvc** [comment]
@ojw28 : is there any ways in v2 to implement this on HLS? I didn't see FormatEvaluator anymore.

**Comment by ojw28** [comment]
All track selection is converged into TrackSelector and TrackSelection in V2, and works in the same way for DASH, SmoothStreaming and HLS. The demo app allows you to select from the available streams in all cases. I suggest you take a look at the trackselection package if you're interested in the internal details.

**Comment by toanvc** [comment]
@ojw28 : Thanks for response! I saw `AdaptiveVideoTrackSelection` and I did it successfully.
I have questions: when I used 
>`DEFAULT_MAX_INITIAL_BITRATE = 800000;`

in my setting as the same value in `AdaptiveVideoTrackSelection`, but it seems always choose the low bitrate from my stream (my stream support the lowest is `2000000` and the highest is `4000000`).
After I changed that value to 

> `DEFAULT_MAX_INITIAL_BITRATE = Integer.MAX_VALUE;`

my player run with adaptive bitrate and played with highest bitrate in fast internet connection.
The question is, can I use `Integer.MAX_VALUE` for my default `maxInitialBitrate` setting? And what's `maxInitialBitrate` exactly does? I'm confused with its' meaning in the comment codes

>`The maximum bitrate in bits per second that should be assumed when a bandwidth estimate is unavailable.`

Thanks!
